ghproxy/ghcache/ghcache.go

```diff
@@ -232,37 +244,116 @@ func NewDiskCache(delegate http.RoundTripper, cacheDir string, cacheSizeGB, maxC
 			maxConcurrency,
 		)
 	}
+
+	go func() {
+		for range time.NewTicker(cachePruneInterval).C {
+			prune(cacheDir)
+		}
+	}()
 	return NewFromCache(delegate,
-		func(partitionKey string) httpcache.Cache {
+		func(partitionKey string, expiresAt time.Time) httpcache.Cache {
+			basePath := path.Join(cacheDir, "data", partitionKey)
+			tempDir := path.Join(cacheDir, "temp", partitionKey)
+			if err := writecachePartitionMetadata(basePath, tempDir, expiresAt); err != nil {
+				logrus.WithError(err).Warn("Failed to write cache metadata file, pruning will not work")
+			}
 			return diskcache.NewWithDiskv(
 				diskv.New(diskv.Options{
-					BasePath:     path.Join(cacheDir, "data", partitionKey),
-					TempDir:      path.Join(cacheDir, "temp", partitionKey),
+					BasePath:     basePath,
+					TempDir:      tempDir,
 					CacheSizeMax: uint64(cacheSizeGB) * uint64(1000000000), // convert G to B
 				}))
 		},
 		maxConcurrency,
 	)
 }
 
+func prune(baseDir string) {
+	// All of this would be easier if the structure was base/partition/{data,temp}
+	// but because of compatibility we can not change it.
+	for _, dir := range []string{"data", "temp"} {
+		base := path.Join(baseDir, dir)
+		cachePartitionCandidates, err := os.ReadDir(base)
+		if err != nil {
+			logrus.WithError(err).Warn("os.ReadDir failed")
+			// no continue, os.ReadDir returns partial results if it encounters an error
+		}
+		for _, cachePartitionCandidate := range cachePartitionCandidates {
+			if !cachePartitionCandidate.IsDir() {
+				continue
+			}
+			metadataPath := path.Join(base, cachePartitionCandidate.Name(), cachePartitionMetadataFileName)
+
+			// Read optimistically and return errors
```

Comment:
What do you mean by "return errors" ?

Reply:
Should have been "ignore errors", fixed.